### ISSUE TITLE
Number keys 

### ISSUE THREAD

**Issue by unphased** [neovim/neovim#2204]
Sorry if this has been addressed in docs, but I searched and couldn't find anything. 

in my vimrc I have 

```
set <k0>=^[Op
set <k1>=^[Oq
set <k2>=^[Or
set <k3>=^[Os
set <k4>=^[Ot
set <k5>=^[Ou
set <k6>=^[Ov
set <k7>=^[Ow
set <k8>=^[Ox
set <k9>=^[Oy
```

This is mainly being used with vim-mark now (which already has binds for the numpad `<k0-9>` keys). 

However I am finding that in neovim, the `set <k0>=` isn't actually successfully setting a keycode (`E846: Key code not set: <k0>`) like it does in Vim. 

It also appears that the numpad numbers are being interpreted the same as the actual numbers, so it adds to the ephemeral number count buffer. Which is fine, but not the behavior I am looking for.

So I am wondering what the proper way to do this is. Has setting keycodes with `set` been abolished?

**Comment by justinmk** [comment]
> `set <k0>=` isn't actually successfully setting a keycode 

Did you try setting it to something other than a literal byte sequence? `^[Op` is (presumably) a terminal control sequence, this probably won't work in Neovim, and there's no point to it.

**Comment by unphased** [comment]
What would I set it to? I don't know how else to have it actually function in vanilla Vim.

**Comment by unphased** [comment]
Can you tell me what you mean by `set <k0>=foo`? What would such a thing accomplish? That is not a sane terminal escape code. I tried it, and afterwards neovim still gives E846 when I ask for what `<k0>` is set to.

**Comment by justinmk** [comment]
> foo is not a sane terminal escape code

Right. There is no reason to set anything to a terminal escape code in neovim.

**Comment by unphased** [comment]
Okay, so what's the verdict? neovim does not support arbitrary terminal codes?

Where does it fetch the escape codes to interpret then? terminfo?

**Comment by justinmk** [comment]
What happens if you do this:

```
:map <k0> :echo 'hi'<cr>
:call feedkeys("\<k0>")
```

**Comment by unphased** [comment]
It printed `hi`, just like it should. That doesn't have any bearing on me being able to drive `<k0>` with my computer keyboard (or remote computer keyboard over SSH through tmux, which is how I really use it most of the time, but which also isn't relevant)

**Comment by justinmk** [comment]
Does `0` on your numpad echo "hi" or not?

**Comment by unphased** [comment]
No, it continues to function identically to the 0 key.

**Comment by justinmk** [comment]
@unphased It may require some extra logic in the TUI. Seems similar to the CTRL-H vs backspace issue https://github.com/neovim/neovim/pull/1820#issuecomment-73033723

**Comment by unphased** [comment]
I did actually sort out a kind of solution for the Ctrl+H vs backspace (0x8) differentiation, basically neovim's TUI is going above and beyond and confusing itself by reading the terminfo def (which most other systems have been ignoring). 

Based on this clue I went and tested libtermkey's `demo` program and it shows that libtermkey is properly detecting the keypad numbers.

**Comment by unphased** [comment]
I can only assume that this code here 

https://github.com/neovim/neovim/blob/0bce4dc54427d05ab320a88f6269a9c1b05ea899/src/nvim/getchar.c#L1430

is related

**Comment by unphased** [comment]
Can somebody comment on how to make these keys map to <k0> thru <k9> in neovim (this works fine in vim)?

From `nvim/getchar.c:1430`

```
      case K_K0:          c = '0'; break;
      case K_K1:          c = '1'; break;
      case K_K2:          c = '2'; break;
      case K_K3:          c = '3'; break;
      case K_K4:          c = '4'; break;
      case K_K5:          c = '5'; break;
      case K_K6:          c = '6'; break;
      case K_K7:          c = '7'; break;
      case K_K8:          c = '8'; break;
      case K_K9:          c = '9'; break;
```

**Comment by justinmk** [comment]
> It would seem that Neovim is going out of its way to interpret `<k0>` as 0, `<k1>` as 1

@unphased That part of getchar.c translates a keycode to the non-mapped result. If control reaches that point, then I'm confused. Are you certain that `case K_K9` is executed (add a printf statement perhaps)? 

>  I went and tested libtermkey's demo program and it shows that libtermkey is properly detecting the keypad numbers. 

It may be a `tui.c` bug then.

> `set <k0>=` isn't actually successfully setting a keycode

Confirmed.

**Comment by justinmk** [comment]
I am guessing that `set <k0>={sequence}` is not covered by any tests and probably a lookup was factored out because someone did not expect such a feature to be used. I'm not so sure this is a feature we necessarily want to support. Fiddling with terminal sequences is something we would hope users don't need to do...

**Comment by unphased** [comment]
Well sure but it's pretty clear that nvim is flat out refusing to fire <kx> keys so that my vimscripts cannot differentiate the keypresses! So it is just that it's alarming that the *workaround* is also not functioning...

**Comment by justinmk** [comment]
Sure, we should support keypad input. I may have a PR up in the next 30 minutes, if you want to try it.

**Comment by justinmk** [comment]
Well, my idea didn't work out, I'll post it tomorrow. In the meantime to answer your question, look at `tk_getkeys`.

**Comment by justinmk** [comment]
I've narrowed it down to at least this fact: the following line in `tk_getkeys` reports keypad input 9 as `9`:

    tk_getkey(input->tk, &key, force)

**Comment by unphased** [comment]
Hey thanks for looking into the issue!

Sounds like you are really close to something. But either way, that is a helpful starting point.

**Comment by justinmk** [comment]
@unphased Does `tput smkx` change anything for you regarding this issue?

See also https://github.com/neovim/neovim/issues/3211 and http://stackoverflow.com/q/21175561/152142.

**Comment by talw** [comment]
Anything new regarding this issue?

**Comment by justinmk** [comment]
No.

**Comment by justinmk** [comment]
#9793 f6fb370b1bcc (Nvim 0.4.0) adds support for keypad keys.

**Comment by unphased** [comment]
I am really glad that numeric keypad is going to work soon, but I would really still like to have some ability to tell neovim that I want to reassign a key to an arbitrary escape code. This has always been possible to do in vim with set <key>=^[code, but this is impossible to do in neovim without apparently customizing a terminfo, unless I have misunderstood something. That's a limitation.

**Comment by unphased** [comment]
Hi, I'm trying to get my custom key code to work and all I've found is the same thread that I authored. Whether or not numpad keys work I don't really care much these days because I kinda stopped using full keyboards (easier to reach mouse is more valuable). 

However it is a pretty big problem for neovim to force me to only ever use "standard" ANSI escape key codes, when vim supports it out of the box with `set`.

**Comment by unphased** [comment]
Hi `<evil grin>` I'm back!

Now I am aiming to add a completely new escape sequence for something that has no semantic equivalent. I'm trying to use `\E[Y` and `\E[2Y` (skipping over Z since `\E[Z` is a defacto shift+tab!) 

As we already hashed out in this issue here neovim has no feature to specify a new escape code to interpret, no way to `set <key>=escapecode` like Vim has. 

As a result unfortunately it looks like if I want to achieve what I want I am going to have to switch to a non utilized existing key of which there are many (like the high-function keys). This is okay and I will go on to do just that this evening but I'm just posting here to complain? I guess? that this situation has rather poor ergonomics.

**Comment by justinmk** [comment]
> without introducing more complexity in the way of defining new terminfo entries

Can also be done with your terminal's config, right (not terminfo)?

> I am going to have to resort to changing the key code sent into vim to a non utilized existing key

Would it not work to use a single key like `<F32>` as the "leader", and then you can have any sequence following that, to define "unlimited" different custom inputs?

**Comment by Mikilio** [comment]
My issue may not have the same objective as the original issue here, but since I have seen @justinmk solve a similar issue and he has answered pretty recently I thougt of taking it here. I use a layered keyboard with dedicated cut, copy and paste buttons. According to [this](https://github.com/torvalds/linux/blob/master/include/uapi/linux/input-event-codes.h) this correspnds to keycodes 137, 133, 135. However I feel like those keycodes have not much in common with the ones used in neovim so in don't know how to bind them as they haven't been given a name yet.

Maybe that issue is being tracked somewhere else already.
Since to me it means improving the ability to add custom keycodes, I added it here.
I apologize if it was uncalled for, but I didn't want to unnecessarily increase the number of issues.

**Comment by zeertzjq** [comment]
That sounds like somethings that should be requested to terminal emulators first.

**Comment by Mikilio** [comment]
Mine does support those keycodes though. Using wezterm.

**Comment by zeertzjq** [comment]
> Mine does support those keycodes though. Using wezterm.
> 
> how they relate to my OS's keycodes

They don't.

**Comment by Mikilio** [comment]
> They don't.

That explains a lot.

**Comment by unphased** [comment]
> > without introducing more complexity in the way of defining new terminfo entries
> 
> Can also be done with your terminal's config, right (not terminfo)?

> > I am going to have to resort to changing the key code sent into vim to a non utilized existing key
> 
> Would it not work to use a single key like `<F32>` as the "leader", and then you can have any sequence following that, to define "unlimited" different custom inputs?

Yeah, defining a custom leader is fine, I haven't been wanting to introduce more than one leader key in my own setup yet but it will happen soon and lead to a large change as i try to make a mnemonic that conforms to it.

Sooo there's a bit of nuance that I glazed over and will attempt to explain better. 

Vim supports `set <F32>=\E[anythingyouwant`. 
Neovim doesn't support that.

However the reality of it is, neovim not supporting this doesn't really have too substantive an impact. 

- Even though vim can be configured easily to understand arbitrary escape sequences, the actual utilization of those key codes is still gonna be confusing ("the hell is F32 anyway" as you review your init down the line... you'll have to look up what escape code F32 was set to, and then look up what this escape code's semantics are supposed to be). It would be a different story if you could actually define your own key names!
- Yep, Neovim won't let you set F32 to any escape code you want, it's only going to respond to F32 via whatever it is (likely `\E[19;5~`) which is just as well, because choosing a poor escape sequence (probably) will lead to even harder to debug situations if you make something conflicting with any existing escape sequences. At the end of the day, you're still gonna need to look up what F32 maps to while reviewing your config. 
- Due to neovim using what so far looks like a consistent naming scheme with tmux, this actually helps a lot to restore sanity, like I confirmed with my current round of improvements, I can set `C-M-S-F11` in tmux and it sends `\E[23;8~` which just shows up properly in nvim as `C-M-S-F11`. This is a function key with so many modifiers tacked onto it that it is way down the [list](https://invisible-island.net/xterm/xterm-function-keys.html) and probably would appear as like F95 or something 😅 The point with this is at least I can read my tmux config, then read my nvim config, and easily see how this ctrl meta shift f11 key is used to 

In an ideal world I would define some new semantic key and a corresponding escape code for it, and configure stuff to understand that, well, actually though this is exactly what terminfo is... and without being hardcore enough to carefully craft those and critique how well neovim and tmux and whatever else is going to interpret it... I'm ok with just letting tmux be the source of truth. What I'm using right now to implement a contextual behavior down this whole stack is: 

- Alacritty config (one for linux, one for macos) translates from os-specific `Ctrl+PgUp` or `Cmd+Shift+[` into `\E[2Y`, an invented escape code. I would give it a semantic name like PrevTab
- tmux is configured with
  ```
set -s user-keys[8] "\e[2Y"
bind -n User8 if -F '#{||:#{m/r:^N?VIM,#{pane_title}},#{==:1,#{session_windows}}}' 'send-keys M-C-S-F11' previous-window
  ```
  This translates further into `Alt+Ctrl+Shift+F11` which was arbitrarily chosen as something I'll never seriously manually use, 
- but which is still accessible from nvim:
  ```
vim.keymap.set('n', "<M-C-S-F11>", "<cmd>lua CycleWindowsOrBuffers(false)<cr>", { desc = "Cycle Windows or Buffers Reverse" })
  ```
  (`CycleWindowsOrBuffers()` is my [own lua function](https://github.com/unphased/vim-config/blob/31e2b03cdf5b09a6446c14f5754aea0a2b96dc28/nvim/init.lua#L2060) so you don't need to go hunting for it on google)

Now if i were more enterprising or pedantic or anal, I would perhaps go and try to establish some sort of new terminfo key capability or something, I don't know how that stuff works, call it TabPrev, `kTP` or `kTPrev`, assign it to `\E[2Y`, and assuming neovim and tmux and whatever else can be made to interpret it, I might be able to reference `<TabPrev>` from nvim and `TabPrev` from tmux and things will be dandy. But I'm not going to do that because I have better things to do with my time than learn terminfo/termcap/etc. 

Since I originated this issue I'm willing to close it if that's what this amounts to.
